### ISSUE TITLE
[VZ] CF Growing Pains

### ISSUE THREAD

**Issue by contolini** [cfpb/capital-framework#196]
![Growing Pains](https://cloud.githubusercontent.com/assets/1060248/10665655/a563f642-789a-11e5-8583-3ef920836a1e.jpg)

CF is rapidly growing (hooray!) and teams are using it (huge win!). Buuuut the bigger it gets, the more time it takes to maintain it. Two years ago [this week](http://GHE/flapjack/front-end/issues/30) we decided to architect CF components as separate repos. [Modularization and composability FTW](https://en.wikipedia.org/wiki/Unix_philosophy). I think we made two mistakes with this decision:

1. We conflated modularity and compartmentalization
  - I kiiiind of regret breaking out CF into separate repos. It can be modular without being siloed into separate repos. Every time we update CF's testing tools or docs layout, we have to roll out those changes to 20 different repos. It's a pain. It's *not* a pain because there's redundant source code, it's a pain because we didn't foresee how frequently we'd iterate on components' testing, packaging and documentation. It's also annoying to have to constantly navigate in and out of directories when hacking on CF.
1. We didn't give enough forethought to CF maintenance.
  - Manually updating 20 modules' package.json's sucks (for example, if we want to switch from pa11y to node-wcag for 508 testing). Other projects handle this with a custom [CLI](https://github.com/lodash/lodash-cli) or a centralized test script that's stored in its own repo but these things risk creating even more technical debt. We currently do it all manually which is not fun.

**Note:** I'm still advocating for every component to have its own manifest with dependency listing and semantic versioning. I'm simply suggesting that we move all the components' source code into the main capital-framework repo and use a build script to test and publish the components.

I don't 100% know what this would look like. Maybe we'd keep a `package.tmpl` file that gets processed for every component as a [prepublish script](https://docs.npmjs.com/misc/scripts)? Maybe we'd just move every component into a `components` sub-directory of capital-framework? Maybe we'd have a separate `npm` (or `bower`) branch that stores everything in master's `src` directory after some processing?

I'll hack on a POC over the next couple weeks. In the meantime, please share any thoughts.

**TL;DR:** The larger CF grows, the harder it is to maintain. We need to fix this. I propose we move all source code into a single repo.

**Comment by jimmynotjim** [comment]
Another issue I'm seeing is now that we're releasing more often, PRs quickly become stale due to either the author or the reviewer not having the same exact versions of another module. I'd hate to peg versions, because a release to one module means updating every other modules dependencies. At the same time, having the ability to install a component that's not released, like say my breakpoints branch on my own fork, is a really great way to test in-progress work on a real project. It's going to be a fine line to walk, but I'm definitely in-favor of a single source that is somehow broken up when published.

**Comment by Scotchester** [comment]
@contolini Definitely interested in you pursuing this concept and developing a POC.

@jimmynotjim If we don't peg dependencies, what's the plan for preventing accidental breakage?

And you bring up a good point about testing unreleased component WIPs. How would we pull that off in this kind of scenario? I'd like to avoid any copy-and-paste-based solution. I can maybe see some ways to handle it with extensive branching.

**Comment by cfarm** [comment]
yesssssssss one repo one repo

**Comment by KimberlyMunoz** [comment]
Concept posted in https://github.com/cfpb/capital-framework/issues/210

**Comment by contolini** [comment]
For the curious, here's my [WIP branch](https://github.com/contolini/capital-framework/tree/voltrazord).

**Comment by Scotchester** [comment]
I believe this is closed now that Voltrazord is done. Feel free to re-open if I missed something.
